README.md
[![Published on NPM](https://img.shields.io/npm/v/@advanced-rest-client/paper-fab-menu.svg)](https://www.npmjs.com/package/@advanced-rest-client/paper-fab-menu)

[![Build Status](https://travis-ci.org/advanced-rest-client/paper-fab-menu.svg?branch=stage)](https://travis-ci.org/advanced-rest-client/paper-fab-menu)

[![Published on webcomponents.org](https://img.shields.io/badge/webcomponents.org-published-blue.svg)](https://www.webcomponents.org/element/@advanced-rest-client/paper-fab-menu)

# <paper-fab-menu>

Material design:
[Floating Action Button](https://www.google.com/design/spec/components/buttons-floating-action-button.html)

A floating action button represents the primary action in an application.

Use the `<paper-fab-menu>` to display menu-like fab buttons and to promote actions.

### Example

```html
<paper-fab-menu icon="add">
  <paper-fab mini title="Favorites" icon="star"></paper-fab>
  <paper-fab mini title="Refresh" icon="refresh"></paper-fab>
  <paper-fab mini title="Text label" label="D"></paper-fab>
</paper-fab-menu>
```

### In a LitElement template

```javascript
import { LitElement, html } from 'lit-element';
import '@advanced-rest-client/paper-fab-menu/paper-fab-menu.js';

class SampleElement extends LitElement {
  render() {
    return html`
    <paper-fab-menu icon="add">
      <paper-fab mini title="Favorites" icon="star"></paper-fab>
      <paper-fab mini title="Refresh" icon="refresh"></paper-fab>
      <paper-fab mini title="Text label" label="D"></paper-fab>
    </paper-fab-menu>
    `;
  }
}
customElements.define('sample-element', SampleElement);
```

### Styling
Style the menu using `<paper-fab>` variables.

The element uses `--paper-fab-menu-background-color` variable to set a
background color or the main `<paper-fab>` element.

### Development

```sh
git clone https://github.com/@advanced-rest-client/paper-fab-menu
cd paper-fab-menu
npm install
```

### Running the demo locally

```sh
npm start
```

### Running the tests
```sh
npm test
```


CHANGELOG.md
<a name="3.0.0"></a>
# [3.0.0](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.5...3.0.0) (2019-01-08)


### Update

* Added new Travis configuration ([8fa90aae250584ad0e17ded91c82a7658a4731d2](https://github.com/advanced-rest-client/paper-fab-menu/commit/8fa90aae250584ad0e17ded91c82a7658a4731d2))
* removed vertical position, fixed an issue where it was possible to click on element when it was hidden ([6d09e394fc5287bb1bdb56ad45e94e4d6302f269](https://github.com/advanced-rest-client/paper-fab-menu/commit/6d09e394fc5287bb1bdb56ad45e94e4d6302f269))



<a name="2.0.6"></a>
## [2.0.6](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.5...v2.0.6) (2017-01-31)


### Update

* Added new Travis configuration ([8fa90aae250584ad0e17ded91c82a7658a4731d2](https://github.com/advanced-rest-client/paper-fab-menu/commit/8fa90aae250584ad0e17ded91c82a7658a4731d2))
* removed vertical position, fixed an issue where it was possible to click on element when it was hidden ([6d09e394fc5287bb1bdb56ad45e94e4d6302f269](https://github.com/advanced-rest-client/paper-fab-menu/commit/6d09e394fc5287bb1bdb56ad45e94e4d6302f269))



<a name="2.0.5"></a>
## [2.0.5](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.4...v2.0.5) (2016-10-06)


### New

* Added dependencyci ([cd8db50d8f1b0bee4a98196e818c3c39c7011e91](https://github.com/advanced-rest-client/paper-fab-menu/commit/cd8db50d8f1b0bee4a98196e818c3c39c7011e91))
* Added hero image ([215a330e7542d768c7b61b2dad5d73270c5fc1c4](https://github.com/advanced-rest-client/paper-fab-menu/commit/215a330e7542d768c7b61b2dad5d73270c5fc1c4))



<a name="2.0.4"></a>
## [2.0.4](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.3...v2.0.4) (2016-09-29)


### Update

* Updated docs ([e8b8071d6611ddadc67f2147eb82804162861724](https://github.com/advanced-rest-client/paper-fab-menu/commit/e8b8071d6611ddadc67f2147eb82804162861724))



<a name="2.0.3"></a>
## 2.0.3 (2016-08-29)

* New: Added Travis integration ([c93f31a](https://github.com/advanced-rest-client/paper-fab-menu/commit/c93f31a))



<a name="2.0.2"></a>
## 2.0.2 (2016-08-29)

* Added automation tasks ([a80b731](https://github.com/advanced-rest-client/paper-fab-menu/commit/a80b731))
* Added changelog ([4c736ba](https://github.com/advanced-rest-client/paper-fab-menu/commit/4c736ba))
* Transfered element to new repository, added tests and demos ([dbe0e1f](https://github.com/advanced-rest-client/paper-fab-menu/commit/dbe0e1f))
* Fix: Fixed linter errors ([c0ff186](https://github.com/advanced-rest-client/paper-fab-menu/commit/c0ff186))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.1...v2.0.1) (2016-08-29)


### Fix

* Fixed linter errors ([c0ff186](https://github.com/advanced-rest-client/paper-fab-menu/commit/c0ff186))
# [3.0.0](https://github.com/advanced-rest-client/paper-fab-menu/compare/2.0.5...3.0.0) (2019-07-17)


### Breaking

* Upgrading to Polymer 3 ([d9e5882c824e5889fbed0728bf1d3464213d08bb](https://github.com/advanced-rest-client/paper-fab-menu/commit/d9e5882c824e5889fbed0728bf1d3464213d08bb))

### build

* bumping version ([bd9c939802f6bb06d39ef986ee1478bf0b9b82e6](https://github.com/advanced-rest-client/paper-fab-menu/commit/bd9c939802f6bb06d39ef986ee1478bf0b9b82e6))

### New

* Adding npm ignore file ([a8e620d4099fddc68310913f3340e9d9eaeef15d](https://github.com/advanced-rest-client/paper-fab-menu/commit/a8e620d4099fddc68310913f3340e9d9eaeef15d))
* Adding test coverage ([94af70dee935f2503490164a99b1e6ed09d86a4e](https://github.com/advanced-rest-client/paper-fab-menu/commit/94af70dee935f2503490164a99b1e6ed09d86a4e))

### refactor

* updating component to use LitElement ([4648ae3d6d7a923823190af97730b15cead7bc4c](https://github.com/advanced-rest-client/paper-fab-menu/commit/4648ae3d6d7a923823190af97730b15cead7bc4c))

### test

* removing endge deom tests because of some wired behavior ([2b615c076a860a38270dea25f8976416d85a1695](https://github.com/advanced-rest-client/paper-fab-menu/commit/2b615c076a860a38270dea25f8976416d85a1695))
* updating test ame in SL ([b0427f51b8c5da280d5957d720ed7aa5de0f298d](https://github.com/advanced-rest-client/paper-fab-menu/commit/b0427f51b8c5da280d5957d720ed7aa5de0f298d))

### Update

* Added new Travis configuration ([8fa90aae250584ad0e17ded91c82a7658a4731d2](https://github.com/advanced-rest-client/paper-fab-menu/commit/8fa90aae250584ad0e17ded91c82a7658a4731d2))
* removed vertical position, fixed an issue where it was possible to click on element when it was hidden ([6d09e394fc5287bb1bdb56ad45e94e4d6302f269](https://github.com/advanced-rest-client/paper-fab-menu/commit/6d09e394fc5287bb1bdb56ad45e94e4d6302f269))
* Removing `@apply` from styles ([a9bcf8fa8ed88aff67b7496361c51481b8971e80](https://github.com/advanced-rest-client/paper-fab-menu/commit/a9bcf8fa8ed88aff67b7496361c51481b8971e80))
* Removing old files ([2697718be5e5cd75a0fc7f4a377f95562cae9b2a](https://github.com/advanced-rest-client/paper-fab-menu/commit/2697718be5e5cd75a0fc7f4a377f95562cae9b2a))
* Updating paths to match web spec ([a6c1ab286edc89571a0787eb1a84a528fab5748d](https://github.com/advanced-rest-client/paper-fab-menu/commit/a6c1ab286edc89571a0787eb1a84a528fab5748d))
* Upgrading dependencies ([61fd0e3b0c1752165f230e7934d7b6ced74b3db8](https://github.com/advanced-rest-client/paper-fab-menu/commit/61fd0e3b0c1752165f230e7934d7b6ced74b3db8))





test/paper-fab-menu.test.js
import { fixture, assert, aTimeout } from '@open-wc/testing';
import sinon from 'sinon/pkg/sinon-esm.js';
import { a11ySuite } from '@advanced-rest-client/a11y-suite/index.js';
import '../paper-fab-menu.js';

describe('<paper-fab-menu>', function() {
  async function basicFixture() {
    // Note, keep text node in this fixture (empty text node)
    return (await fixture(`<paper-fab-menu icon="add">
    <paper-fab mini title="Polymer" icon="polymer"></paper-fab>
    <paper-fab mini title="Favorites" icon="star"></paper-fab>

    </paper-fab-menu>`));
  }

  async function noChildrenFixture() {
    return (await fixture(`<paper-fab-menu icon="add"></paper-fab-menu>`));
  }

  describe('Basics', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Children are hidden', () => {
      const nodes = element.querySelectorAll('paper-fab');
      assert.equal(nodes[0].className.indexOf('opened'), -1);
    });

    it('Children are initially processed', () => {
      const nodes = element.querySelectorAll('paper-fab');
      assert.equal(nodes[0].getAttribute('role'), 'menuitem');
    });

    it('Added child is processed', async () => {
      const div = document.createElement('div');
      element.appendChild(div);
      await aTimeout(120);
      assert.equal(div.getAttribute('role'), 'menuitem');
      assert.equal(div.getAttribute('aria-hidden'), 'true');
    });
  });

  describe('toggle()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Toggles "opened" state', () => {
      element.toggle();
      assert.isTrue(element.opened);
    });
  });

  describe('_testOpen()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Clears __closingDebouncer', () => {
      element.opened = true; // it's just for coverage
      element.__closingDebouncer = 123;
      element._testOpen();
      assert.isUndefined(element.__closingDebouncer);
    });

    it('Clears timeout', () => {
      element.__closingDebouncer = 123;
      const spy = sinon.spy(window, 'clearTimeout');
      element._testOpen();
      window.clearTimeout.restore();
      assert.isTrue(spy.called);
    });

    it('Opens the menu', () => {
      element._testOpen();
      assert.isTrue(element.opened);
    });
  });

  describe('_testClose()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Sets __closingDebouncer', (done) => {
      element.opened = true;
      element._testClose();
      assert.typeOf(element.__closingDebouncer, 'number');
      setTimeout(() => done(), 20);
    });

    it('Does nothing when __closingDebouncer is already set', () => {
      element.__closingDebouncer = 1;
      element._testClose();
      assert.equal(element.__closingDebouncer, 1);
    });

    it('Closes the menu', async () => {
      element.opened = true;
      element._testClose();
      await aTimeout();
      assert.isFalse(element.opened);
    });

    it('Clears __closingDebouncer', async () => {
      element.opened = true;
      element._testClose();
      await aTimeout();
      assert.isUndefined(element.__closingDebouncer);
    });
  });

  describe('_detectClick()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Opens the element when the click is on the element', () => {
      element.click();
      assert.isTrue(element.opened);
    });

    it('Closes the element when opened and click is outside the element', () => {
      element.opened = true;
      document.body.click();
      assert.isFalse(element.opened);
    });

    it('Does nothing when not opened (outside click)', () => {
      document.body.click();
      assert.isUndefined(element.opened);
    });

    it('Does nothing when opened (element click)', () => {
      element.opened = true;
      element.click();
      assert.isTrue(element.opened);
    });

    it('Does nothing when clicked on child element', () => {
      element.opened = true;
      element.querySelector('paper-fab').click();
      assert.isTrue(element.opened);
    });
  });

  describe('_nodesAddedHandler()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Does nothing when no argument', () => {
      // Coverage
      element._nodesAddedHandler();
    });

    it('Does nothing when empty argument', () => {
      // Coverage
      element._nodesAddedHandler([]);
    });

    it('Adds "role" attribute to a node', () => {
      const node = document.createElement('div');
      element._nodesAddedHandler([node]);
      assert.equal(node.getAttribute('role'), 'menuitem');
    });

    it('Adds "aria-hidden" attribute to a node', () => {
      const node = document.createElement('div');
      element._nodesAddedHandler([node]);
      assert.equal(node.getAttribute('aria-hidden'), 'true');
    });

    it('Ignores non-element nodes', () => {
      const node = document.createElement('div');
      const txt = document.createTextNode('test');
      element._nodesAddedHandler([txt, node]);
      // no error on setting attribute on text node.
    });
  });

  describe('_processInitialNodes()', () => {
    let element;
    beforeEach(async () => {
      element = await basicFixture();
    });

    it('Nodes added declaratively have role attribute', () => {
      const node = element.querySelector('paper-fab');
      assert.equal(node.getAttribute('role'), 'menuitem');
    });

    it('Nodes added declaratively have aria-hidden attribute', () => {
      const node = element.querySelector('paper-fab');
      assert.equal(node.getAttribute('aria-hidden'), 'true');
    });
  });

  describe('_updateDelay()', () => {
    let element;
    let children;
    beforeEach(async () => {
      element = await basicFixture();
      await aTimeout();
      children = element.querySelectorAll('paper-fab');
    });

    it('Sets transitionDelay on each child', () => {
      element._updateDelay(children);
      assert.equal(children[0].style.transitionDelay, '25ms');
      assert.equal(children[1].style.transitionDelay, '75ms');
    });

    it('Adds "opened" class when opened', () => {
      element._updateDelay(children, true);
      assert.isTrue(children[0].classList.contains('opened'));
      assert.isTrue(children[1].classList.contains('opened'));
    });

    it('Updates aria-hidden attribute', () => {
      element._updateDelay(children, true);
      assert.equal(children[0].getAttribute('aria-hidden'), 'false');
      assert.equal(children[1].getAttribute('aria-hidden'), 'false');
    });

    it('Returns delay time', () => {
      const result = element._updateDelay(children, true);
      assert.equal(result, 125);
    });

    it('Removes opened class when closed state', () => {
      children[0].classList.add('opened');
      children[1].classList.add('opened');

      element._updateDelay(children, false);
      assert.isFalse(children[0].classList.contains('opened'));
      assert.isFalse(children[1].classList.contains('opened'));
    });

    it('Updates aria-hidden attribute when closed state', () => {
      children[0].setAttribute('aria-hidden', 'false');
      children[1].setAttribute('aria-hidden', 'false');

      element._updateDelay(children, false);
      assert.equal(children[0].getAttribute('aria-hidden'), 'true');
      assert.equal(children[1].getAttribute('aria-hidden'), 'true');
    });
  });

  describe('_openedTimeout()', () => {
    let element;
    beforeEach(async () => {
      element = await noChildrenFixture();
    });

    it('Sets childrenVisible to false when closed', async () => {
      element.opened = true;
      await aTimeout();
      element._openedTimeout([], false);
      await aTimeout(25);
      assert.isFalse(element.childrenVisible);
    });

    it('Keeps childrenVisible when opened', async () => {
      element.childrenVisible = true;
      element._openedTimeout([], true);
      await aTimeout(25);
      assert.isTrue(element.childrenVisible);
    });
  });

  describe('a11y', () => {
    a11ySuite('Normal state', `<paper-fab-menu icon="add">
    <paper-fab mini title="Polymer" icon="polymer"></paper-fab>
    <paper-fab mini title="Favorites" icon="star"></paper-fab>
    </paper-fab-menu>`);

    a11ySuite('Opened state', `<paper-fab-menu icon="add" opened>
    <paper-fab mini title="Polymer" icon="polymer"></paper-fab>
    <paper-fab mini title="Favorites" icon="star"></paper-fab>
    </paper-fab-menu>`);
  });
});


tasks/release.js
var gulp = require('gulp');
var conventionalGithubReleaser = require('conventional-github-releaser');
var conventionalChangelog = require('gulp-conventional-changelog');
var bump = require('gulp-bump');
var gutil = require('gulp-util');
var git = require('gulp-git');
var fs = require('fs');
var runSequence = require('run-sequence');

gulp.task('changelog', function() {
  return gulp.src('CHANGELOG.md', {
      buffer: false
    })
    .pipe(conventionalChangelog({
      preset: 'auto' // Or to any other commit message convention you use.
    }))
    .pipe(gulp.dest('./'));
});

gulp.task('github-release', function(done) {
  conventionalGithubReleaser({
    type: 'oauth',
    // change this to your own GitHub token or use an environment variable
    token: process.env.GITHUB_TOKEN
  }, {
    preset: 'angular' // Or to any other commit message convention you use.
  }, done);
});
gulp.task('bump-version', function() {
  // We hardcode the version change type to 'patch' but it may be a good idea to
  // use minimist (https://www.npmjs.com/package/minimist) to determine with a
  // command argument whether you are doing a 'major', 'minor' or a 'patch' change.
  return gulp.src(['./bower.json', './package.json'])
    .pipe(bump({
      type: 'patch'
    }).on('error', gutil.log))
    .pipe(gulp.dest('./'));
});
gulp.task('commit-changes', function() {
  return gulp.src('.')
    .pipe(git.add())
    .pipe(git.commit('[Prerelease] Bumped version number'));
});

gulp.task('push-changes', function(cb) {
  git.push('origin', 'master', cb);
});

function getPackageJsonVersion() {
  // We parse the json file instead of using require because require caches
  // multiple calls so the version number won't be updated
  return JSON.parse(fs.readFileSync('./package.json', 'utf8')).version;
}

gulp.task('create-new-tag', function(cb) {
  var version = getPackageJsonVersion();
  git.tag(version, 'Created Tag for version: ' + version, {signed: true}, function(error) {
    if (error) {
      return cb(error);
    }
    git.push('origin', 'master', {
      args: '--tags'
    }, cb);
  });
});

gulp.task('release', function(callback) {
  if (!process.env.GITHUB_TOKEN) {
    callback(new Error('The GITHUB_TOKEN environment variable is not set.'));
    return;
  }
  runSequence(
    'bump-version',
    'changelog',
    'commit-changes',
    'push-changes',
    'create-new-tag',
    'github-release',
    function(error) {
      if (error) {
        console.log(error.message);
      } else {
        console.log('RELEASE FINISHED SUCCESSFULLY');
      }
      callback(error);
    });
});


karma.sl.config.js
/* eslint-disable import/no-extraneous-dependencies */
const merge = require('webpack-merge');
const slSettings = require('@advanced-rest-client/testing-karma-sl/sl-settings.js');
const createBaseConfig = require('./karma.conf.js');

module.exports = (config) => {
  const cnf = slSettings();
  cnf.sauceLabs.testName = 'connectivity-state';
  cnf.browsers = [
    'SL_Chrome',
    'SL_Chrome-1',
    'SL_Firefox',
    'SL_Firefox-1',
    'SL_Safari',
    'SL_Safari-1',
    // 'SL_EDGE',
  ];
  if (process.env.TRAVIS) {
    const buildLabel =
      'TRAVIS #' + process.env.TRAVIS_BUILD_NUMBER + ' (' + process.env.TRAVIS_BUILD_ID + ')';

    cnf.browserStack = {
      build: buildLabel,
      tunnelIdentifier: process.env.TRAVIS_JOB_NUMBER,
    };

    cnf.sauceLabs.build = buildLabel;
    cnf.sauceLabs.tunnelIdentifier = process.env.TRAVIS_JOB_NUMBER;
  }

  config.set(merge(createBaseConfig(config), cnf));

  return config;
};


paper-fab-menu.d.ts
/**
 * DO NOT EDIT
 *
 * This file was automatically generated by
 *   https://github.com/Polymer/tools/tree/master/packages/gen-typescript-declarations
 *
 * To modify these typings, edit the source file(s):
 *   paper-fab-menu.js
 */


// tslint:disable:variable-name Describing an API that's defined elsewhere.
// tslint:disable:no-any describes the API as best we are able today

import {LitElement, html, css} from 'lit-element';

declare namespace UiElements {

  /**
   * Material design:
   * [Floating Action Button](https://www.google.com/design/spec/components/buttons-floating-action-button.html)
   *
   * A floating action button represents the primary action in an application.
   *
   * Use the `<paper-fab-menu>` to display menu-like fab buttons and to promote actions.
   *
   * ### Example
   *
   * ```html
   * <paper-fab-menu color="teal" icon="add">
   *  <paper-fab mini icon="star"></paper-fab-menu-item>
   *  <paper-fab mini icon="star"></paper-fab-menu-item>
   *  <paper-fab mini icon="star"></paper-fab-menu-item>
   * </paper-fab-menu>
   * ```
   *
   * ### Styling
   * Style the menu using `<paper-fab>` variables.
   *
   * The element uses `--paper-fab-menu-background-color` variable to set a
   * background color or the main `<paper-fab>` element.
   */
  class PaperFabMenu extends LitElement {

    /**
     * True when the menu is opened.
     */
    opened: boolean|null|undefined;

    /**
     * The icon to render. It's binded to `paper-fab`'s icon property.
     */
    icon: string|null|undefined;

    /**
     * If true then the children container is displayed.
     */
    childrenVisible: boolean|null|undefined;
    constructor();
    connectedCallback(): void;
    disconnectedCallback(): void;
    firstUpdated(): void;
    render(): any;
    _childrenUpdated(mutations: any): void;
    _setChanged(prop: any, value: any): any;

    /**
     * Toogles the menu
     */
    toggle(): void;

    /**
     * Tests if the menu should be opened and opens it if nescesary.
     */
    _testOpen(): void;

    /**
     * Tests if the menu should be closed and closes it if nescesary
     */
    _testClose(): void;

    /**
     * Renders the mennu opened / closed
     *
     * @param opened Current menu state
     */
    _openedChanged(opened: Boolean|null): void;
    _openedTimeout(children: any, opened: any): void;

    /**
     * Updates animation delay time attribute in distributed children.
     *
     * @returns Time of the latest delay set on the children animations.
     */
    _updateDelay(children: Array<Element|null>|null, opened: Boolean|null): Number|null;

    /**
     * Closes menu when click is detected outside this control.
     */
    _detectClick(e: MouseEvent|null): void;

    /**
     * MutationObserver initialized in the constructor does not
     * triggers changes when the element is initialized. This
     * function processes nodes set up declaratively when the element is still
     * initializing.
     */
    _processInitialNodes(): void;

    /**
     * Adds `role` and `aria-hidden` attribbutes to passed nodes.
     *
     * @param nodes List of nodes to process.
     */
    _nodesAddedHandler(nodes: Array<Node|null>|null): void;
  }
}

declare global {

  interface HTMLElementTagNameMap {
    "paper-fab-menu": UiElements.PaperFabMenu;
  }
}


paper-fab-menu-item.js
/**
@license
Copyright 2018 Pawel Psztyc, The ARC team

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import {PolymerElement} from '../../@polymer/polymer/polymer-element.js';
import {html} from '../../@polymer/polymer/lib/utils/html-tag.js';
import '../../@polymer/paper-fab/paper-fab.js';
/**
 * The `<paper-fab-menu-item>` is a menu item that can be used with `<paper-fab-menu>`.
 *
 * It renders as a mini fab button but it's hidden. The `<paper-fab-menu>` triggers menu items when hovered.
 *
 * See `<paper-fab-menu>` for more information and demos.
 * @customElement
 * @polymer
 * @demo demo/index.html
 * @memberof UiElements
 */
class PaperFabMenuItem extends PolymerElement {
  static get template() {
    return html`
    <style>
    :host {
      display: block;
      flex: 1;
      flex-basis: 0.000000001px;
      margin: 4px;
    }

    .menu-item {
      display: block;
      position: relative;
      transition: 300ms all cubic-bezier(0.165, 0.84, 0.44, 1);
      transform: scale(0);
    }

    .menu-item.opened {
      transform: scale(1);
    }
    </style>
    <div class\$="[[_computeMainClass(opened)]]">
      <paper-fab mini="" icon="[[icon]]" label="[[label]]"></paper-fab>
    </div>
`;
  }

  static get properties() {
    return {
      // Color of the menu item.
      color: {
        type: String,
        observer: '_colorChanged'
      },
      // The delay (in ms) before the entry / exit animation begins.
      transitionDelay: {
        type: Number,
        observer: '_updateTransitionDelay'
      },
      // Label only used with paper-fab when the icon is not present.
      label: String,
      // The icon to render. It's binded to `paper-fab`'s icon property.
      icon: String,
      /**
       * If true this item will be rendered as opened.
       * This property is used by the `paper-fab-menu` element to determine
       * control state. Do not use this property.
       */
      opened: {
        type: Boolean,
        value: false
      }
    };
  }

  _updateTransitionDelay(transitionDelay) {
    this.shadowRoot.querySelector('.menu-item').style.transitionDelay = transitionDelay + 'ms';
  }

  // Called when color has changed.
  _colorChanged(color) {
    this.updateStyles({
      '--paper-fab-background': color
    });
    // For some reason code above do not affect main fab at bootstrap.
    this.shadowRoot.querySelector('paper-fab').style.backgroundColor = color;
  }

  _computeMainClass(opened) {
    let clazz = 'menu-item';
    if (opened) {
      clazz += ' opened';
    }
    return clazz;
  }
}
window.customElements.define('paper-fab-menu-item', PaperFabMenuItem);


paper-fab-menu.js
/**
@license
Copyright 2018 Pawel Psztyc, The ARC team

Licensed under the Apache License, Version 2.0 (the "License"); you may not
use this file except in compliance with the License. You may obtain a copy of
the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
License for the specific language governing permissions and limitations under
the License.
*/
import { LitElement, html, css } from 'lit-element';
import '@polymer/paper-fab/paper-fab.js';
/**
 * Material design:
 * [Floating Action Button](https://www.google.com/design/spec/components/buttons-floating-action-button.html)
 *
 * A floating action button represents the primary action in an application.
 *
 * Use the `<paper-fab-menu>` to display menu-like fab buttons and to promote actions.
 *
 * ### Example
 *
 * ```html
 * <paper-fab-menu color="teal" icon="add">
 *  <paper-fab mini icon="star"></paper-fab-menu-item>
 *  <paper-fab mini icon="star"></paper-fab-menu-item>
 *  <paper-fab mini icon="star"></paper-fab-menu-item>
 * </paper-fab-menu>
 * ```
 *
 * ### Styling
 * Style the menu using `<paper-fab>` variables.
 *
 * The element uses `--paper-fab-menu-background-color` variable to set a
 * background color or the main `<paper-fab>` element.
 *
 * @customElement
 * @demo demo/index.html
 * @memberof UiElements
 */
class PaperFabMenu extends LitElement {
  static get styles() {
    return css`:host {
      width: 60px;
    }

    .main-items ::slotted(a) {
      display: block;
      text-decoration: none;
    }

    .main-items ::slotted(*) {
      margin: 4px;
      transition: 200ms transform cubic-bezier(0.165, 0.84, 0.44, 1);
      transform: scale(0);
    }

    .main-items ::slotted(.opened) {
      transform: scale(1) !important;
    }

    .main-items {
      display: flex;
      flex-direction: column;
      flex: 1;
      flex-basis: 0.000000001px;
      align-self: center;
      flex-basis: 0%;
    }

    .main-items.fix-safari {
      flex-basis: auto;
    }

    .menu-fab-button {
      display: flex;
      flex-direction: column;
      align-self: center;
    }

    .paper-fab-menu-container {
      display: flex;
      flex-direction: column;
    }

    :host(:not([childrenvisible])) .main-items {
      display: none;
    }

    paper-fab {
      background-color: var(--paper-fab-menu-background-color, var(--accent-color));
      transition: 200ms transform;
      transform : rotate(0deg);
    }

    :host([opened]) paper-fab {
      transform : rotate(45deg);
    }`;
  }

  render() {
    const { icon, opened } = this;
    return html`<div class="paper-fab-menu-container">
      <div class="main-items">
        <slot id="items" slot="content"></slot>
      </div>
      <div class="menu-fab-button">
        <paper-fab tabindex="0" .icon="${icon}" ?opened="${opened}"></paper-fab>
      </div>
    </div>`;
  }

  static get properties() {
    return {
      // True when the menu is opened.
      opened: { type: Boolean, reflect: true },
      // The icon to render. It's binded to `paper-fab`'s icon property.
      icon: String,
      // If true then the children container is displayed.
      childrenVisible: {
        type: Boolean,
        reflect: true
      }
    };
  }

  get opened() {
    return this._opened;
  }

  set opened(value) {
    if (this._setChanged('opened', value)) {
      this._openedChanged(value);
      // Polymer copatibility
      this.dispatchEvent(new CustomEvent('opened-changed', { detail: { value } }));
    }
  }

  constructor() {
    super();
    this._detectClick = this._detectClick.bind(this);
    this._testOpen = this._testOpen.bind(this);
    this._testClose = this._testClose.bind(this);

    const config = { attributes: false, childList: true, subtree: false };
    this._observer = new MutationObserver((mutations) => this._childrenUpdated(mutations));
    this._observer.observe(this, config);
  }

  connectedCallback() {
    super.connectedCallback();
    this.addEventListener('mouseover', this._testOpen);
    this.addEventListener('mouseout', this._testClose);
    document.body.addEventListener('click', this._detectClick);

    if (!this.hasAttribute('role')) {
      this.setAttribute('role', 'menubar');
    }
    this.setAttribute('aria-haspopup', 'true');
  }

  disconnectedCallback() {
    super.disconnectedCallback();
    this.removeEventListener('mouseover', this._testOpen);
    this.removeEventListener('mouseout', this._testClose);
    document.body.removeEventListener('click', this._detectClick);
  }

  firstUpdated() {
    this._processInitialNodes();
    this._openedChanged(this.opened);
  }

  _childrenUpdated(mutations) {
    for (const mutation of mutations) {
      if (mutation.addedNodes.length) {
        this._nodesAddedHandler(mutation.addedNodes);
      }
    }
  }

  _setChanged(prop, value) {
    const key = `_${prop}`;
    const old = this[key];
    if (value === old) {
      return false;
    }
    this[key] = value;
    if (this.requestUpdate) {
      this.requestUpdate(prop, old);
    }
    return true;
  }

  // Toogles the menu
  toggle() {
    this.opened = !this.opened;
  }

  /**
   * Tests if the menu should be opened and opens it if nescesary.
   */
  _testOpen() {
    if (this.__closingDebouncer) {
      clearTimeout(this.__closingDebouncer);
      this.__closingDebouncer = undefined;
    }
    if (!this.opened) {
      this.opened = true;
    }
  }

  /**
   * Tests if the menu should be closed and closes it if nescesary
   */
  _testClose() {
    if (this.__closingDebouncer) {
      return;
    }
    this.__closingDebouncer = setTimeout(() => {
      this.opened = false;
      this.__closingDebouncer = undefined;
    });
  }

  /**
   * Renders the mennu opened / closed
   * @param {Boolean} opened Current menu state
   */
  _openedChanged(opened) {
    const slot = this.shadowRoot && this.shadowRoot.querySelector('slot');
    if (!slot) {
      return;
    }
    const children = Array.from(slot.assignedNodes());
    if (opened) {
      children.reverse();
      this.childrenVisible = true;
    }
    setTimeout(() => this._openedTimeout(children, opened));
  }

  _openedTimeout(children, opened) {
    const delayTime = this._updateDelay(children, opened);
    if (!opened) {
      setTimeout(() => {
        this.childrenVisible = false;
      }, delayTime);
    }
  }
  /**
   * Updates animation delay time attribute in distributed children.
   * @param {Array<Element>} children
   * @param {Boolean} opened
   * @return {Number} Time of the latest delay set on the children animations.
   */
  _updateDelay(children, opened) {
    let time = 25;
    for (let i = 0, len = children.length; i < len; i++) {
      const child = children[i];
      if (child.nodeType !== 1) {
        continue;
      }
      child.style.transitionDelay = time + 'ms';
      if (opened) {
        child.classList.add('opened');
        child.setAttribute('aria-hidden', 'false');
      } else {
        child.classList.remove('opened');
        child.setAttribute('aria-hidden', 'true');
      }
      time += 50;
    }
    return time;
  }
  /**
   * Closes menu when click is detected outside this control.
   * @param {MouseEvent} e
   */
  _detectClick(e) {
    let path = e.path;
    if (!path && e.composedPath) {
      path = e.composedPath();
    }
    if (!path || !path.length) {
      return;
    }
    let isMainButton = false;
    for (let i = 0, len = path.length; i < len; i++) {
      if (path[i] === this) {
        isMainButton = true;
      }
    }
    if (isMainButton) {
      if (!this.opened) {
        this.opened = true;
      }
      return;
    }
    if (this.opened) {
      this.opened = false;
    }
  }
  /**
   * MutationObserver initialized in the constructor does not
   * triggers changes when the element is initialized. This
   * function processes nodes set up declaratively when the element is still
   * initializing.
   */
  _processInitialNodes() {
    const slot = this.shadowRoot && this.shadowRoot.querySelector('slot');
    const children = slot.assignedNodes();
    this._nodesAddedHandler(children);
  }
  /**
   * Adds `role` and `aria-hidden` attribbutes to passed nodes.
   * @param {Array<Node>} nodes List of nodes to process.
   */
  _nodesAddedHandler(nodes) {
    if (!nodes || !nodes.length) {
      return;
    }
    for (let i = 0, len = nodes.length; i < len; i++) {
      const node = nodes[i];
      if (node.nodeType !== 1) {
        continue;
      }
      node.setAttribute('role', 'menuitem');
      node.setAttribute('aria-hidden', 'true');
    }
  }
}
window.customElements.define('paper-fab-menu', PaperFabMenu);
